toopher/toopher_iframe.py

```diff
@@ -0,0 +1,191 @@
+                return False
+        except Exception as e:
+            logging.error(e)
+            return False
+
+    def _urldecode_iframe_data(self, data):
+        data_dict = urlparse.parse_qs(data['toopher_iframe_data'])
+        return dict((k,v[0]) for (k,v) in data_dict.items())
+
+    def _validate_data(self, data, request_token, kwargs):
+
+        self._check_for_missing_keys(data)
+        self._verify_session_token(data.get('session_token'), request_token)
+        self._check_if_signature_is_expired(data.get('timestamp'), kwargs)
+        self._validate_signature(data)
+        return data
+
+    def _check_for_missing_keys(self, data):
+        missing_keys = []
+        for required_key in ('toopher_sig', 'timestamp', 'session_token'):
+            if not required_key in data:
+                missing_keys.append(required_key)
+
+        if missing_keys:
+            raise SignatureValidationError('Missing required keys: {0}'.format(', '.join(missing_keys)))
+
+    def _verify_session_token(self, session_token, request_token):
+        if request_token:
+            if request_token != session_token:
+                raise SignatureValidationError('Session token does not match expected value!')
+
+    def _check_if_signature_is_expired(self, timestamp, kwargs):
+        ttl = kwargs.pop('ttl') if 'ttl' in kwargs else DEFAULT_IFRAME_TTL
+        ttl_valid = int(time.time()) - int(timestamp) < ttl
+        if not ttl_valid:
+            raise SignatureValidationError('TTL expired')
+
+    def _validate_signature(self, data):
+        maybe_sig = data['toopher_sig']
+        del data['toopher_sig']
+
+        try:
+            computed_signature = self._calculate_signature(data)
+        except Exception as e:
+            raise SignatureValidationError('Error while calculating signature: {0}'.format(e.args[0]))
+
+        if not maybe_sig == computed_signature:
+            raise SignatureValidationError('Computed signature does not match submitted signature: {0} vs {1}'.format(computed_signature, maybe_sig))
+
+    def _create_authentication_request_dict(self, data):
+        return {
+            'id': data['id'],
+            'pending': True if data['pending'] == 'true' else False,
+            'granted': True if data['granted'] == 'true' else False,
+            'automated': True if data['automated'] == 'true' else False,
+            'reason': data['reason'],
+            'reason_code': data['reason_code'],
+            'terminal': {
+                'id': data['terminal_id'],
+                'name': data['terminal_name'],
+                'requester_specified_id': data['terminal_requester_specified_id'],
+                'user': {
+                    'id': data['pairing_user_id'],
+                    'name': data['user_name'],
+                    'toopher_authentication_enabled': True if data['user_toopher_authentication_enabled'] == 'true' else False
```

Comment:
maybe simplify this to just `data['user_toopher_authentication_granted'] == 'true'`